src/main/java/io/hops/hopsworks/expat/migrations/featurestore/featureview/CreateFeatureViewFromTrainingDataset.java

```diff
@@ -0,0 +1,381 @@
+
+        Integer[] features = getArray("training_dataset_feature", trainingDatasetId);
+
+        insertFeatureViewStatement.setString(1, name);
+        insertFeatureViewStatement.setInt(2, featurestoreId);
+        insertFeatureViewStatement.setTimestamp(3, created);
+        insertFeatureViewStatement.setInt(4, creator);
+        insertFeatureViewStatement.setInt(5, version);
+        insertFeatureViewStatement.setString(6, description);
+        insertFeatureViewStatement.setString(7, label);
+        // inode
+        setInode(insertFeatureViewStatement, projectName, userName, name, version);
+        if (!dryRun) {
+          int affectedRows = insertFeatureViewStatement.executeUpdate();
+          // need to commit here, otherwise tables are locked.
+          connection.commit();
+          if (affectedRows != 1) {
+            throw new MigrationException("Creating feature view failed, no rows affected.");
+          } else {
+            ResultSet generatedKeys = insertFeatureViewStatement.getGeneratedKeys();
+            if (generatedKeys.next()) {
+              Integer fvId = generatedKeys.getInt(1);
+              setFeatureView("training_dataset_feature", "training_dataset", trainingDatasetId, fvId);
+              setFeatureView("training_dataset_filter", "training_dataset_id", trainingDatasetId, fvId);
+              setFeatureView("training_dataset_join", "training_dataset", trainingDatasetId, fvId);
+              setFeatureView("training_dataset", "id", trainingDatasetId, fvId);
+              connection.commit();
+              setXAttr(featurestoreId, getFeatureViewFullPath(projectName, name, version).toString(),
+                  description, created, email, features);
+            } else {
+              throw new MigrationException("Creating feature view failed, no ID obtained.");
+            }
+          }
+        }
+      }
+      insertFeatureViewStatement.close();
+      getTrainingDatasetsStatement.close();
+      connection.commit();
+      connection.setAutoCommit(true);
+      LOGGER.info(n + " training dataset records have been updated.");
+    } catch (SQLException e) {
+      throw new MigrationException("Migration failed. Cannot commit.", e);
+    } finally {
+      super.close();
+    }
+  }
+
+  public void runRollback() throws RollbackException {
+    // 1. remove fv from td, td feature, td join, td filter
+    // 2. remove file and inode
+    // 3. remove fv
+    try {
+      Integer n = 0;
+      connection.setAutoCommit(false);
+      PreparedStatement getFeatureViewsStatement = connection.prepareStatement(GET_ALL_FEATURE_VIEWS);
+      ResultSet featureViews = getFeatureViewsStatement.executeQuery();
+      Set<String> projectNames = Sets.newHashSet();
+      while (featureViews.next()) {
+        n += 1;
+        Integer fvId = featureViews.getInt("id");
+        projectNames.add(featureViews.getString("projectname"));
+
+        if (!dryRun) {
+          removeFeatureViewFromTables(fvId);
```

Comment:
does this actually leave the training datasets in a usable state for the old api?

also, we are not moving training datasets into the new .featureview directory, right?

Reply:
> does this actually leave the training datasets in a usable state for the old api?

yes, we just add the fv id to all tables of training dataset during migration and do not alter other thing.

> also, we are not moving training datasets into the new .featureview directory, right?

no, Training datasets are stored in the original path as the old API.